add options for limited and full random milestones and awards

vsrisuknimit opened the pull request:
Before clicking random Ms/As
![image](https://user-images.githubusercontent.com/14239220/96321471-8ea36d80-0fe3-11eb-9b07-9c1f25b84f68.png)
After turning on random Ms/As, two options are provided.
![image](https://user-images.githubusercontent.com/14239220/96321482-9cf18980-0fe3-11eb-86b0-0e8fc1aa7ea2.png)

Options are:
- Limited synergy as proposed in #1311 (slight change, max total = 20 instead of 15, number of high synergy = 2 instead of 1)
- Full random (does not care at all about synergy)

The distribution of the limited synergy option is as follow:
![image](https://user-images.githubusercontent.com/14239220/96321571-e7730600-0fe3-11eb-8620-13abb96bbc6a.png)

vsrisuknimit commented:
@Lynesth I rewrote the algorithm, I believe it is much cleaner now.

@chosta I probably screw up some vue things. If you could take a look, it would be nice. The function seems to randomize properly. I also want the flex items ("Limited Synergy" and "Full Random") to span the whole row. Not sure how to do that.

vsrisuknimit commented:
@nwai90 I would like to add the following to the wiki. https://github.com/bafolts/terraforming-mars/wiki/Variants#random-milestones-and-awards (limited synergy subsection)

**Limited Synergy Milestones and Awards Randomization**
Only allows combination of milestones and awards that follow the conditions listed below. Synergy scores are based on this [synergy matrix](https://docs.google.com/spreadsheets/d/1PIUzO9XLF-MHSrTDP9QU2VW3ce7hJNciAR-G7N-RBUA/edit#gid=330638583).
1) Total synergy score is 20 or lower. (Tharsis Board has total synergy score around 20. The other two boards have much lower scores.)
2) No pair with synergy score 7 or above. (Prohibit landlord & desert settler or terraformer & benefactor.)
3) No more than two pairs that are 4+ in synergy score. (Allow two mayor & landlord or gardener & polar explorer.)

nwai90 commented:
Updated the wiki page

chosta commented:
@vsrisuknimit Vue is fine. I'd suggest, when you have double option on a single line to use the full width of the column. Have the two elements with same width and center the content. Let me know if that makes sense. This is done best with `flex` and `justify-content` etc. I can help.

vsrisuknimit commented:
The box stretch to full column width now.

![image](https://user-images.githubusercontent.com/14239220/96335235-47e65f80-1045-11eb-950a-42f718adebe1.png)

vsrisuknimit commented:
With help from @Lynesth 

![image](https://user-images.githubusercontent.com/14239220/96337668-b2ec6200-1056-11eb-9789-6c6ece86705b.png)

vsrisuknimit commented:
This PR can close #1621 and #1311 .

Tagging @alrusdi @nwai90 for review.

kberg commented on src/Game.ts:
```diff
@@ -55,10 +55,12 @@ import { IParty } from "./turmoil/parties/IParty";
 import { LogHelper } from "./components/LogHelper";
 import { ColonyName } from "./colonies/ColonyName";
-import { getRandomMilestonesAndAwards } from "./MilestoneAwardSelector";
+import { getRandomMA_OneByOne } from "./MilestoneAwardSelector";
```
Please rename this -- In the PR where getRandomMilestonesAndAwards was introduced, it was to get away from unclear terms like MASynergy. Not to mention that _ isn't a standard methodName.

vsrisuknimit replied:
renamed

kberg commented on server.ts:
```diff
@@ -443,6 +443,7 @@ function createGame(req: http.IncomingMessage, res: http.ServerResponse): void {
                 startingCorporations: gameReq.startingCorporations,
                 shuffleMapOption: gameReq.shuffleMapOption,
                 randomMA: gameReq.randomMA,
+                randomMAOption: gameReq.randomMAOption,
```
There's no need for randomMA and randomMAOption. Also, having randomMAOption as a string means it's hard to figure out how to track down the valid values. Please replace it with an enum. In fact, consider this

```
enum RandomMilestoneAward {
  NONE = "none,
  LIMITED = "limited,
  UNLIMITED = "unlimited"
}

randomMA: RandomMilestoneAward

vsrisuknimit replied:
changed to enum

kberg commented on src/MilestoneAwardSelector.ts:
```diff
@@ -206,3 +206,106 @@ export function computeSynergy(indexes: Array<number>) : number {
+export function getRandomMA_OneByOne(withVenusian: boolean = true, 
+    numberMARequested: number, 
+    maxSynergyAllowed: number = 6, 
+    totalSynergyAllowed: number = 20,
```
Thanks for including this after my comment, but I don't think we need both this and maxSynergyAllowed. No need to overcomplicate it.

kberg replied:
In fact, remove most of this -- focus on one single meaningful improvement and end it there. Less to test. We can make more changes later.

vsrisuknimit replied:
Disagree. I believe this is an improvement over simple max synergy for all pairs.

kberg replied:
I believe you that it is -- an improvement at the cost of complexity is my concern. I see two ares where this will be a problem.

1. The test cases to ensure correctness.
2. What controls to hand off to the end-user. I was suggesting somewhere (discord?) a slider from 0-9 or 0-50 (those numbers aren't quite perfect but close) and presenting all four of these as options are too much. Someone shouldn't have to know how the innards work to make a smart selection.

none/limited/unlimited is a good start, but given that everyone's aware of the analysis _right now_ I'd rather settle on the end-game first.

vsrisuknimit replied:
The algorithm is not all complex. It should actually be faster than the previous method which generate a whole set at once and keep rejecting them until one pass.
1. We do need test cases. Not sure what we are testing though. Randomize correctly without duplicate? Randomize correctly with the parameters given?
2. Slider can be implemented on top of this if need be. I think it is better to give user limited options here and then a filter option for fully customizing which milestones and awards to choose. No need for users to play around with sliders and understanding how the algorithm works.

nwai90 replied:
A finite set of predefined choices presented to the user would be best here imo. At the first iteration users don't have to know every detail that happens under the hood to decide whether to randomize or not.

About the test cases, I would pick one or two specific combinations/pairs that are guaranteed to never show up with synergy restriction applied. Given a large enough number of X runs, we should expect:

(1) to never see these combinations if synergy option is selected, and
(2) to see them occurring at least once in the case of unlimited synergy

kberg replied:
> The algorithm is not all complex.  It should actually be faster than the previous method which generate a whole set at once and keep rejecting them until one pass.

I assert that you and I think it is not complex because we've bothered to read the code and figure it out. Don't add additional complexity, please. Gamers aren't going to need them, what's the meaningful extra value? I'm glad they'll be faster though speed isn't a concern.

kberg reviewed (changes requested):
Definitely needs tests once the general API and structure are settled.

vsrisuknimit commented:
I want to have Limited selected when Randomizer is on. The value of v-model is set to limited right now. Just not sure how to check the box.
![image](https://user-images.githubusercontent.com/14239220/96354009-8ad52100-109f-11eb-9b00-138d00042ac0.png)

vsrisuknimit commented:
If the randomizer requires more than 5 attempts to get a satisfying set of Ms and As that follow the rules, an error is thrown to prevent infinite recursion loop.

Below is a histogram of the number of attempt required to generate a set of Ms and As that follow the rules. [6,20,2,4] for [max, total, numberOfPair, highThreshold].

![image](https://user-images.githubusercontent.com/14239220/96378016-545add00-1157-11eb-8f9a-f77fb50286cb.png)

vsrisuknimit commented:
Tagging @chosta again for help with pre-checking the limited synergy radio button as soon as Random MA checkbox is checked.

vsrisuknimit commented:
- Added constants
- Added function to verify the synergy rules
- Added tests to MilestonesAwardsSelector.spec.ts.
- Added checks to make sure no infinite loop can happen.

vsrisuknimit commented:
I believe this is good to go.

Tagging @nwai90 again for review.

kberg commented:
I chatted with @vsrisuknimit and cleared up my misunderstanding. Looks great. Thank you!

samyelias commented:
> If the randomizer requires more than 5 attempts to get a satisfying set of Ms and As that follow the rules, an error is thrown to prevent infinite recursion loop.

I just noticed this. If you remove ineligible goals as you pick them, you'll never get an infinite loop and the randomizer will run a little faster. This is how I implemented it on my randomizer (I zero out the weights as it picks, since I use weights).

kberg commented:
No matter what, I think 5 tries is a little too short. It was never that
slow to begin with -- why make such a small count?

>

vsrisuknimit commented:
> No matter what, I think 5 tries is a little too short. It was never that slow to begin with -- why make such a small count?
> […](#)

because of this: https://github.com/bafolts/terraforming-mars/pull/1637#issuecomment-711411034

but we could increase to 10 if you want.

kberg commented:
I stand corrected! You have DATA! A comment in the code pointing to that
would be awesome.

>
